sys/riscv/riscv/pmap.c

```diff
@@ -3603,6 +3639,464 @@ pmap_enter_quick_locked(pmap_t pmap, vm_offset_t va, vm_page_t m,
+
+	m = PHYS_TO_VM_PAGE(PTE_TO_PHYS(oldpte));
+	if (*mp == m) {
+		/*
+		 * We expected this page here (i.e., this is the page we just
+		 * scanned), so go ahead and update.  We know that the CLG bits
+		 * must still be wrong in light of the earlier test.
+		 */
+		res = PMAP_CAPLOADGEN_OK;
+
+		if (!(flags & PMAP_CAPLOADGEN_HASCAPS)) {
+			/*
+			 * We didn't see a capability on this page; step this
+			 * PTE closer to being cap-clean.
+			 */
+			if (oldpte & PTE_CD) {
+				/*
+				 * CAP-DIRTY -> DIRTYABLE; we just do a store
+				 * here as there might be a capability store
+				 * between us having examined the page and the
+				 * store-release below, and so we're going to
+				 * scan the page again anyway.
+				 */
+				pmap_clear_bits(pte, PTE_CD);
+			} else if (oldpte & PTE_CW) {
+				/* PTE CAP-DIRTYABLE -> CAP-CLEAN? */
+				pmap_caploadgen_update_clear_cw(pte, oldpte);
+			} else if (flags & PMAP_CAPLOADGEN_NONEWMAPS) {
+				/* No new mappings possible */
+				vm_page_astate_t mas = vm_page_astate_load(m);
+
+				if (mas.flags & PGA_CAPDIRTY) {
+					/* Page-level cleaning: -?> VACANT */
+					vm_page_aflag_clear(m, PGA_CAPDIRTY);
+				} else {
+					KASSERT(mas.flags & PGA_CAPSTORE,
+					    ("Page already CAP-CLEAN?"));
+					/* PTE CAP-CLEAN; page -?> IDLE */
+
+					/* XXXNWF Why? */
+					if (flags & PMAP_CAPLOADGEN_UPDATETLB) {
+						sfence_vma_page(va);
+					}
+					PMAP_UNLOCK(pmap);
+					/*
+					 * XXXMJ don't we need to hold a
+					 * reference to m?  What prevents
+					 * concurrent reclamation?
+					 */
+					pmap_caploadgen_test_all_clean(m);
+					m = NULL;
+					goto out_unlocked;
+				}
+			}
+		} else {
+			/*
+			 * Page has caps; may as well mark it dirty if we're
+			 * allowed to store here.
+			 *
+			 * We could clear PGA_CAPDIRTY here, too, but it
+			 * probably doesn't get set often ough to merit.
+			 */
+			if ((oldpte & PTE_CW) && !(oldpte & PTE_CD)) {
+				pmap_store_bits(pte, PTE_CD);
+			}
+		}
+
+		/*
+		 * On the fast path, where we're just updating the CRG bit, this
+		 * is the only store to the PTE we'll do.  This is probably a
+		 * better state of affairs (1 + epsilon AMOs, no retries) than
+		 * we could otherwise easily get (either 1 LL/SC CAS + epsilon
+		 * retries or 1 AMOSWAP to a zero PTE + 1 store).
+		 */
+		pmap_caploadgen_update_crg(pmap, pte);
+
+		if (flags & PMAP_CAPLOADGEN_UPDATETLB) {
+			sfence_vma_page(va);
+		}
+		m = NULL;
+	} else if (!(vm_page_astate_load(m).flags & PGA_CAPSTORE)) {
+		KASSERT(!(oldpte & PTE_CW), ("!PGA_CAPSTORE but CW?"));
+		KASSERT(!(oldpte & PTE_CD), ("!PGA_CAPSTORE but CD?"));
+
+#if defined(INVARIANTS) && defined(CHERI_CAPREVOKE_CLEARTAGS)
+		/*
+		 * We think there aren't any capabilities on this page.  Make
+		 * sure we're not loading one.
+		 *
+		 * In the !CHERI_CAPREVOKE_CLEARTAGS case, we might consider a
+		 * page clean if it contains only data and revoked capabilities,
+		 * so this check is not valid.
+		 */
```

Comment:
Now that we're using `cheri_revoke_is_revoked`, is this still correct only for the `CLEARTAGS` case?